docs/guides/server/reverseproxy.adoc

```diff
@@ -170,6 +170,11 @@ Client certificate lookup via a proxy header for X.509 authentication is conside
 * If passthrough is not an option, implement the following security measures:
 ** Configure your network so that {project_name} is isolated and can accept connections only from the proxy.
 ** Make sure that the proxy overwrites the header that is configured in `spi-x509cert-lookup-<provider>-ssl-client-cert` option.
+** As an additional layer of security, the header name used should be difficult to guess and differ from common default example `SSL_CLIENT_CERT`, rather `SECRET_HEADER_NAME_FOR_SSL_CLIENT_CERT`
+   This can mitigate the case, where the proxy will not reliably overwrite user provided header.
```

Comment:
When would the proxy not do this? Are you thinking of a scenario where TLS is setup to have the client cert as optional? Even then it seems like the proxy would overwrite the header.

If we don't have a concrete example, then this amounts to suggesting security through obscurity.

Reply:
The Apache TLS proxy described in my example serves several backends.
They are using different ProxyPass settings, where Keycloak gets connected via
```
ProxyPass “/auth/” “http://keycloakserver:8080/auth”
ProxyPassReverse “/auth” “http://keycloakserver:8080/auth”
```
In Keycloak several realms are given and are using user/password auth,
but one needs to use x.509 client certificate based login with the usage of `SSL_CLIENT_CERT` header extracted from client TLS handshake in proxy.
The mTLS is not optional but **required** and forced for that special realm URL in proxy with `SSLVerifyClient require`.
Keycloak special realm match:
```
<LocationMatch “/auth/realms/special_x509_realmname/protocol/openid-connect/auth”> 
    SSLVerifyClient require
    SSLOptions +ExportCertData +StrictRequire
    SSLVerifyDeth 3

    RequestHeader set SSL_CLIENT_CERT "%{SSL_CLIENT_CERT}s"
    RequestHeader set CERT_CHAIN_0 "%{SSL_CLIENT_CERT_CHAIN_0}s"
</LocationMatch>
```
Keycloak common match:
```
<LocationMatch “/auth”>
    ......
</LocationMatch>
```

If one does not `RequestHeader unset SSL_CLIENT_CERT` very early before routing to any backend, keycloak gets the attacker injected header due to bypassing the `special_x509_realmname` location with usage of **;foo=bar** in url.
For Apache, attacker url
`/auth/realms/special_x509_realmname/protocol/openid-connect;foo=bar/auth`
is not matching the LocationMatch:
`/auth/realms/special_x509_realmname/protocol/openid-connect/auth`
means Apache takes next LocationMatch “/auth” where no mTLS is required, nor header is unset.

If you don't want to add this config example it's ok, but I repeat that the current documentation is still too vague precisely for this attack case.
Do you believe that there are only superior admin out there who are awesome and make no mistakes? I would never have managed this attack, but it was possible for me.

I agree that the use of special headernames alone only provides security through obscurity. 
However, due to the standard examples with the use of `SSL_CLIENT_CERT`, which any admin will also use during the first setup, a different header name makes the attack more difficult and <ins>could</ins> help.

Reply:
For me, changing the header name **may** help.
I'm not suggesting that it's the perfect solution, but more than one sentence should be spent on this topic to mitigate it.


>If one were to instead use a less specific location match, including having all of the special_x509_realmname traffic with the ssl / header configuration shown or set / unset the SSL_CLIENT_CERT otherwise, there wouldn't be a problem.

If you mean the case where location match would be `/auth/realms/special_x509_realmname`:
In this case, all traffic is mTLS protected, including the internal backend calls.
In my scenario, there are some internal systems that are connected to keycloak via network routing to the proxy entry point for this particular realm.
These systems would need to use a machine-client certificate to overcome the TLS handshake with a required client certificate.

But that wouldn't matter, because an attacker can simply move the `;foo=bar` parameter injection to a different position in URL, in order to trick Apache (and maybe nginx)

`<LocationMatch “/auth/realms/special_x509_realmname”>` is bypassed for example by `/auth/realms/special_x509_realmname;foo=bar/protocol/openid-connect/auth`
`/auth/realms;foo=bar/special_x509_realmname/protocol/openid-connect/auth`
`/auth;foo=bar/realms/special_x509_realmname/protocol/openid-connect/auth`